src/components/number/number.test.jsx

```diff
@@ -0,0 +1,188 @@
+      expect(screen.getByLabelText('Test Component')).toBeRequired();
+      expect(screen.getByLabelText('Test Component')).toBeInvalid();
+    });
+
+    it('can be unset', () => {
+      render(<Number {...requiredProps} required={false} />);
+      expect(screen.getByLabelText('Test Component')).not.toBeRequired();
+      expect(screen.getByLabelText('Test Component')).toBeValid();
+    });
+
+    it('is invalid on mount but does not have an error message', () => {
+      render(<Number {...requiredProps} required />);
+      expect(screen.getByLabelText('Test Component')).toBeInvalid();
+      expect(screen.getByLabelText('Test Component')).toHaveDescription('');
+    });
+
+    it('is invalid after typing', async () => {
+      render(<Number {...requiredProps} required={true} />);
+      userEvent.type(screen.getByLabelText('Test Component'), '2');
+      expect(screen.getByLabelText('Test Component')).toHaveValue(2);
+      expect(screen.getByLabelText('Test Component')).toBeValid();
+      expect(screen.getByLabelText('Test Component')).toHaveDescription('');
+      userEvent.type(screen.getByLabelText('Test Component'), '{backspace}');
+      expect(screen.getByLabelText('Test Component')).toHaveValue(null);
+      expect(screen.getByLabelText('Test Component')).toBeInvalid();
+      expect(screen.getByLabelText('Test Component')).toHaveDescription('Constraints not satisfied');
+    });
+
+    it('is invalid after tabbing through', async () => {
+      render(<Number {...requiredProps} required={true} />);
+      userEvent.tab();
+      expect(document.activeElement).toBe(screen.getByLabelText('Test Component'));
+      expect(screen.getByLabelText('Test Component')).toBeInvalid();
+      expect(screen.getByLabelText('Test Component')).toHaveDescription('');
+      userEvent.tab();
+      expect(document.activeElement).not.toBe(screen.getByLabelText('Test Component'));
+      expect(screen.getByLabelText('Test Component')).toBeInvalid();
+      expect(screen.getByLabelText('Test Component')).toHaveDescription('Constraints not satisfied');
+    });
+  });
+
+  describe('value API', () => {
+    it('sets the value', () => {
+      render(<Number {...requiredProps} value={12} />);
+      expect(screen.getByLabelText('Test Component')).toHaveValue(12);
+    });
+
+    it('updates the value', () => {
+      const { rerender } = render(<Number {...requiredProps} value={12} />);
+      rerender(<Number {...requiredProps} value={21} />);
+      expect(screen.getByLabelText('Test Component')).toHaveValue(21);
+    });
+
+    it('set on the ref', () => {
+      const ref = createRef();
+      render(<Number {...requiredProps} ref={ref} value={10101} />);
+      expect(ref.current.value).toBe(10101);
+      userEvent.type(screen.getByLabelText('Test Component'), '2');
+      expect(ref.current.value).toBe(101012);
+    });
+
+    it('indicates if it is invalid on mount', () => {
```

Comment:
Let's add an assertion on the validation message. Something like:

```
expect(...).toHaveDescription(...);
```

It tests that the input and the field control components are synced.